### ISSUE TITLE
[Feature]: Hash the keys in db

### ISSUE THREAD

**Issue by krrishdholakia** [BerriAI/litellm#983]
### The Feature

Don't store the actual keys in the db. Store a hash. 

Optionally - also salt it. 

### Motivation, pitch

user request - "I'll feel comfortable enough (once hashing is enabled) to switch to virtual keys."

**Comment by ghost** [comment]
"I would use sha512 with https://docs.python.org/3/library/hashlib.html
Python documentation
[hashlib — Secure hashes and message digests](https://docs.python.org/3/library/hashlib.html)
Source code: Lib/hashlib.py This module implements a common interface to many different secure hash and message digest algorithms. Included are the FIPS secure hash algorithms SHA1, SHA224, SHA256,...
Image
sha512 will be hardware accelerated on almost every platform too.
and since the API keys are randomly generated to begin with, you don't really need to worry about salting the hashes IMO.
tl;dr use hashlib to hash the API keys with sha512"

**Comment by ghost** [comment]
"when generating a new API key, also generate a random UUID to store in the database."

**Comment by ghost** [comment]
so you want us to accept/generate a uuid for each key in your db, and that way if you need to generate new keys for that uuid, you can?

**Comment by ghost** [comment]
"related, can we hash the API master and user keys instead of storing them cleartext? Or will that cause issues for some use-cases? 
hashing them would also stop the timing attack potential on the user API keys."

**Comment by ghost** [comment]
Choosing against salting, for now - as it would complicate operations like getting `/key/info`. 

For v0, going to begin with the sha256 hash

**Comment by ghost** [comment]
closing as this is now implemented.
